src/hotspot/share/c1/c1_LIRGenerator.cpp

```diff
@@ -530,12 +530,14 @@ void LIRGenerator::arithmetic_op(Bytecodes::Code code, LIR_Opr result, LIR_Opr l
 
         if (right->is_constant()) {
           jint c = right->as_jint();
-          if (c > 0 && is_power_of_2(c)) {
-            // do not need tmp here
-            __ shift_left(left_op, exact_log2(c), result_op);
-            did_strength_reduce = true;
-          } else {
-            did_strength_reduce = strength_reduce_multiply(left_op, c, result_op, tmp_op);
+          if (c > 0 && c < max_jint) {
```

Comment:
This prevents platform specific optimizations for negative `c`. E.g. multiplication by -1 could be strength reduced to a negation.

Reply:
I didn't get it. How this will affect -1 case ? 

I see that this is implementation for `strength_reduce_multiply`: 
```cpp
bool LIRGenerator::strength_reduce_multiply(LIR_Opr left, jint c, LIR_Opr result, LIR_Opr tmp) {
  assert(left != result, "should be different registers");
  if (is_power_of_2(c + 1)) {
    __ shift_left(left, log2i_exact(c + 1), result);
    __ sub(result, left, result);
    return true;
  } else if (is_power_of_2(c - 1)) {
    __ shift_left(left, log2i_exact(c - 1), result);
    __ add(result, left, result);
    return true;
  }
  return false;
}
```
which will return false in case of `-1`.  

Basically even without my change, `c = -1` will set `did_strength_reduce` to `false`.

Reply:
I'm not talking about existing code. I'm talking about possibilities which you prevent. With your code `strength_reduce_multiply` will no longer be called with negative `c` preventing possible optimizations inside of it.

Reply:
I didn't get your point. Some platform owners may want to use `strength_reduce_multiply` for more cases in the future.